barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp

```diff
@@ -60,110 +118,152 @@ class GoblinTranslatorCircuitBuilder : CircuitBuilderBase<arithmetization::Gobli
+        ACCUMULATOR_HIGH_LIMBS_RANGE_CONSTRAINT_1,
+        ACCUMULATOR_HIGH_LIMBS_RANGE_CONSTRAINT_2,
+        ACCUMULATOR_HIGH_LIMBS_RANGE_CONSTRAINT_3,
+        ACCUMULATOR_HIGH_LIMBS_RANGE_CONSTRAINT_4,
+        ACCUMULATOR_HIGH_LIMBS_RANGE_CONSTRAINT_TAIL,
+        QUOTIENT_LOW_BINARY_LIMBS, // Quotient limbs
+        QUOTIENT_HIGH_BINARY_LIMBS,
+        QUOTIENT_LOW_LIMBS_RANGE_CONSTRAIN_0, // Range constraints for quotient
+        QUOTIENT_LOW_LIMBS_RANGE_CONSTRAIN_1,
+        QUOTIENT_LOW_LIMBS_RANGE_CONSTRAIN_2,
+        QUOTIENT_LOW_LIMBS_RANGE_CONSTRAIN_3,
+        QUOTIENT_LOW_LIMBS_RANGE_CONSTRAIN_4,
+        QUOTIENT_LOW_LIMBS_RANGE_CONSTRAIN_TAIL,
+        QUOTIENT_HIGH_LIMBS_RANGE_CONSTRAIN_0,
+        QUOTIENT_HIGH_LIMBS_RANGE_CONSTRAIN_1,
+        QUOTIENT_HIGH_LIMBS_RANGE_CONSTRAIN_2,
+        QUOTIENT_HIGH_LIMBS_RANGE_CONSTRAIN_3,
+        QUOTIENT_HIGH_LIMBS_RANGE_CONSTRAIN_4,
+        QUOTIENT_HIGH_LIMBS_RANGE_CONSTRAIN_TAIL,
         RELATION_WIDE_LIMBS, // Limbs for checking the correctness of  mod 2²⁷² relations. TODO(kesha): add range
                              // constraints
+        RELATION_WIDE_LIMBS_RANGE_CONSTRAINT_0,
+        RELATION_WIDE_LIMBS_RANGE_CONSTRAINT_1,
+        RELATION_WIDE_LIMBS_RANGE_CONSTRAINT_2,
+        RELATION_WIDE_LIMBS_RANGE_CONSTRAINT_3,
+
+        TOTAL_COUNT
 
     };
-    static constexpr size_t MAX_OPERAND = 3;
+
+    // Basic goblin translator has the minicircuit size of 2048, so optimize for that case
+    static constexpr size_t DEFAULT_TRANSLATOR_VM_LENGTH = 2048;
+
+    // Maximum size of a single limb is 68 bits
     static constexpr size_t NUM_LIMB_BITS = 68;
+
+    // For soundness we need to constrain the highest limb so that the whole value is at most 50 bits
+    static constexpr size_t NUM_LAST_LIMB_BITS = Fq::modulus.get_msb() + 1 - 3 * NUM_LIMB_BITS;
+
+    // 128-bit z_1 and z_2 are split into 2 limbs each
     static constexpr size_t NUM_Z_LIMBS = 2;
-    static constexpr size_t MICRO_LIMB_BITS = 12;
-    static constexpr size_t LEFTOVER_CHUNK_BITS = 8;
+
+    // Number of bits in the quotient representation
+    static constexpr size_t NUM_QUOTIENT_BITS = 256;
+
+    // Number of bits in the quotient highest limb
+    static constexpr size_t NUM_LAST_QUOTIENT_LIMB_BITS = 256 - 3 * NUM_LIMB_BITS;
+
+    // Number of bits in Z scalars
+    static constexpr size_t NUM_Z_BITS = 128;
+    // The circuit builder has a default range constraint mechanism that is used throughout. It range cosntrains the
+    // values to < 2¹⁴
+    static constexpr size_t MICRO_LIMB_BITS = 14;
+
+    // Maximum size of a micro limb used for range constraints
+    static constexpr auto MAX_MICRO_LIMB_SIZE = (uint256_t(1) << MICRO_LIMB_BITS) - 1;
+
+    // To range constrain a limb to 68 bits we need 6 limbs: 5 for 14-bit windowed chunks and 1 to range constrain the
+    // highest to a more restrictive range (0 <= a < 2¹⁴ && 0 <= 4*a < 2¹⁴ ) ~ ( 0 <= a < 2¹² )
```

Comment:
The comment makes sense to me but the comparison expressions are confusing. Is there a typo here? I cant make sense of constraing a to 14 bits then also 4a to 14 bits

Reply:
Imagine, you have a range constraint that ensures all values are in range {0,1,2,3,4,5,6,7}, but you want to range constrain to {0,1}. What you do is range constrain a to [0-7] and then range constrain 4*a to [0-7]. Then a can only be {0,1}